### PR TITLE
CIRC-560 Modify CHECK-OUT processor to support CHECK-IN end expired sessions

### PR DESCRIPTION
**Pull request by Svitlana-Zmiivska**
## Purpose
Modify CHECK-OUT processor to support CHECK-IN end expired sessions
Resolves [CIRC-560](https://issues.folio.org/browse/CIRC-560)

## Approach
- modify CHECK-OUT processor to support CHECK-IN end expired sessions;

## Links
[CIRC-560](https://issues.folio.org/browse/CIRC-560)

## Pre-Merge Checklist:
Before merging this PR, please go through the following list and take appropriate actions.

- Does this PR meet or exceed the expected quality standards?
  - [x] Code coverage on new code is 80% or greater
  - [x] Duplications on new code is 3% or less
  - [x] There are no major code smells or security issues
- Does this introduce breaking changes?
  - [ ] Were any API paths or methods changed, added or removed?
  - [ ] Were there any schema changes?
  - [ ] Did any of the interface versions change?
  - [ ] Were permissions changed, added, or removed?
  - [ ] Are there new interface dependencies?
  - [x] There are no breaking changes in this PR.

### CODE REVIEW

**Comment by Svitlana-Zmiivska** [comment]
This story depends on the story [CIRC-559](https://issues.folio.org/browse/CIRC-559). Can be merge only after the pull request [CIRC-559](https://github.com/folio-org/mod-circulation/pull/406)

File: src/main/java/org/folio/circulation/domain/notice/session/PatronActionSessionRepository.java [hunk 1]
```diff
@@ -115,10 +115,13 @@ private JsonObject mapToJson(PatronSessionRecord patronSessionRecord) {
   public CompletableFuture<Result<MultipleRecords<PatronSessionRecord>>> findPatronActionSessions(
     String patronId, PatronActionType actionType, int limit) {
 
-    final Result<CqlQuery> patronIdQuery = exactMatch(PATRON_ID, patronId);
-    final Result<CqlQuery> actionTypeQuery = exactMatch(ACTION_TYPE, actionType.getRepresentation());
+    Result<CqlQuery> cqlQueryResult = exactMatch(PATRON_ID, patronId);
+    if(!PatronActionType.ALL.equals(actionType)) {
```
Comment (marcjohnson-kint) [inline src/main/java/org/folio/circulation/domain/notice/session/PatronActionSessionRepository.java]: I might be tempted to extract a method for this check. Also the whitespace needs amending.

```suggestion
    if (notAllPatronActionType(actionType)) {
```

Comment (marcjohnson-kint) [inline src/main/java/org/folio/circulation/domain/notice/session/PatronActionSessionRepository.java]: I'd be tempted to include whitespace between these lines to improve readability

File: src/main/java/org/folio/circulation/domain/notice/session/PatronActionSessionRepository.java [hunk 2]
```diff
@@ -115,10 +115,13 @@ private JsonObject mapToJson(PatronSessionRecord patronSessionRecord) {
...
+      final Result<CqlQuery> actionTypeQuery = exactMatch(ACTION_TYPE, actionType.getRepresentation());
+      cqlQueryResult = cqlQueryResult.combine(actionTypeQuery, CqlQuery::and);
+    }
 
-    return patronIdQuery.combine(actionTypeQuery, CqlQuery::and)
+    return cqlQueryResult
```
Comment (marcjohnson-kint) [inline src/main/java/org/folio/circulation/domain/notice/session/PatronActionSessionRepository.java]:
```suggestion
    return sessionsQuery
```

File: src/test/java/api/loans/EndExpiredPatronActionSessionTests.java [hunk 3]
```diff
@@ -9,44 +9,79 @@
...
-import io.vertx.core.json.JsonObject;
+import org.folio.circulation.domain.notice.session.PatronActionType;
+import org.folio.circulation.support.http.client.IndividualResource;
 
 public class EndExpiredPatronActionSessionTests extends APITests {
 
-  @Before
-  public void initSession() throws InterruptedException, MalformedURLException, TimeoutException, ExecutionException {
+  @Test
+  public void expiredEndSessionAfterCheckOut()
```
Comment (marcjohnson-kint) [inline src/test/java/api/loans/EndExpiredPatronActionSessionTests.java]: What is an `expired end session`?

Comment (Svitlana-Zmiivska) [reply src/test/java/api/loans/EndExpiredPatronActionSessionTests.java]: this means that we need to end an expired session.  You can find "End session" term in the description of the story [CIRC-560](https://issues.folio.org/browse/CIRC-560)

File: src/main/java/org/folio/circulation/domain/notice/session/PatronActionSessionRepository.java (hunk 2 above)
Comment (Svitlana-Zmiivska) [reply src/main/java/org/folio/circulation/domain/notice/session/PatronActionSessionRepository.java]: renamed

File: src/main/java/org/folio/circulation/domain/notice/session/PatronActionSessionRepository.java (hunk 1 above)
Comment (Svitlana-Zmiivska) [reply src/main/java/org/folio/circulation/domain/notice/session/PatronActionSessionRepository.java]: Done

Comment (Svitlana-Zmiivska) [reply src/main/java/org/folio/circulation/domain/notice/session/PatronActionSessionRepository.java]: Done

File: src/test/java/api/loans/EndExpiredPatronActionSessionTests.java (hunk 3 above)
Comment (marcjohnson-kint) [reply src/test/java/api/loans/EndExpiredPatronActionSessionTests.java]: Please can you describe the process or scenario that this test is demonstrating.

Is it demonstrating that check out notices are sent when an expired session is ended?

Comment (Svitlana-Zmiivska) [reply src/test/java/api/loans/EndExpiredPatronActionSessionTests.java]: This test demonstrates that we delete user session records from user-session table if a session is expired. 
The case `when check out notices are sent if an expired session`  is covered in PatronActionSessionTests class

File: src/main/java/org/folio/circulation/domain/notice/session/PatronActionSessionRepository.java [hunk 4]
```diff
@@ -115,15 +115,23 @@ private JsonObject mapToJson(PatronSessionRecord patronSessionRecord) {
   public CompletableFuture<Result<MultipleRecords<PatronSessionRecord>>> findPatronActionSessions(
     String patronId, PatronActionType actionType, int limit) {
 
-    final Result<CqlQuery> patronIdQuery = exactMatch(PATRON_ID, patronId);
-    final Result<CqlQuery> actionTypeQuery = exactMatch(ACTION_TYPE, actionType.getRepresentation());
+    Result<CqlQuery> sessionsQuery = exactMatch(PATRON_ID, patronId);
 
-    return patronIdQuery.combine(actionTypeQuery, CqlQuery::and)
+    if(notAllPatronActionType(actionType)) {
```
Comment (kostyantyn-kh) [inline src/main/java/org/folio/circulation/domain/notice/session/PatronActionSessionRepository.java]:
```suggestion
    if (notAllPatronActionType(actionType)) {
```

File: src/main/java/org/folio/circulation/domain/notice/session/PatronActionSessionRepository.java [hunk 5]
```diff
@@ -115,15 +115,23 @@ private JsonObject mapToJson(PatronSessionRecord patronSessionRecord) {
...
+      sessionsQuery = sessionsQuery.combine(actionTypeQuery, CqlQuery::and);
+    }
+
+    return sessionsQuery
       .after(query -> findBy(query, limit))
       .thenCompose(r -> r.combineAfter(
         () -> userRepository.getUser(patronId), this::setUserForLoans));
   }
 
+  private boolean notAllPatronActionType(PatronActionType actionType) {
```
Comment (kostyantyn-kh) [inline src/main/java/org/folio/circulation/domain/notice/session/PatronActionSessionRepository.java]:
```suggestion
  private boolean isPatronActionTypeSpecified(PatronActionType actionType) {
```

**Comment by kostyantyn-kh** [comment]
@Svitlana-Zmiivska 
Please add test to reflect situation when patron has two (or more) sessions and only one should be expired by timeout.

**Pull request merged**